File: leetcode/150.evaluate-reverse-polish-notation.py
# @leet start
class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        # NOTE: I didn't actually read what Reverse Polish notation is,
        # I'm just trying to infer it from the examples—crossing my
        # fingers and hoping it works

        # Make functions for the operations
        operations_dict = {
            "+": lambda x, y: x + y,
            "-": lambda x, y: x - y,
            "*": lambda x, y: x * y,
            "/": lambda x, y: int(x / y),
        }

        # Use a stack here
        stack = []

        for token in tokens:
            if token in operations_dict:
                y = stack.pop()
                x = stack.pop()

                res = operations_dict[token](x, y)

                stack.append(res)
            else:
                stack.append(int(token))

        return stack[-1]
# @leet end

File: leetcode/150.evaluate-reverse-polish-notation.ts
// @leet start
function evalRPN(tokens: string[]): number {
  const stack: number[] = [];

  for (const token of tokens) {
    if ("+-*/".includes(token)) {
      const b = stack.pop()!;
      const a = stack.pop()!;

      switch (token) {
        case "+":
          stack.push(a + b);
          break;
        case "-":
          stack.push(a - b);
          break;
        case "*":
          stack.push(a * b);
          break;
        default:
          stack.push(Math.trunc(a / b));
      }
    } else {
      stack.push(Number(token));
    }
  }

  return stack[0];
}
// @leet end
